Lib/pdb.py

```diff
@@ -588,6 +588,36 @@ def _hold_exceptions(self, exceptions):
             self._chained_exceptions = tuple()
             self._chained_exception_index = 0
 
+    def process_cmdqueue(self):
+        """Process commands in the command queue.
+
+        The function returns in two cases:
+            1. the command queue is empty
+            2. a command to continue execution is encountered
+
+        The return value is
+            True - if the user input is expected after processing,
+            False - otherwise.
+        """
+        if not self.cmdqueue:
+            return True
+
+        # It's possible that cmdqueue does not contain any command that resumes
+        # execution, and _cmdloop will stuck in the loop waiting for user input
+        # in that case. To prevent that, we append a sentinel command (step) at
+        # the end of the queue. This makes sure we can break out of the loop.
+        # If the sentinel command is consumed, we know extra user input is
+        # needed. Otherwise we can pop the sentinel command and return.
+        # The side effect of the sentinel command is that we would execute a
+        # set_step(), but that side effect will be overwritten by the actual
+        # resuming command by the user later.
```

Comment:
I don't understand what you mean here about the side effect.

Reply:
Okay so there might be a misunderstanding here. The sentinel  command (`step`) will not "resume execution" of the program. The only thing it does is to break out of the cmdloop. We can't resume the execution implicitly because that's wrong (in another word, we can really execute a `step` command because that would step the program).

The `step` command we append will only run `set_step()` and break out of the cmdloop - that's all we want. We need to get user input now, like the `step` command never executed. However, `set_step()` has side effect - it will write to some variables (to control which line the debugger needs to stop at when it yields control). But those side effects will be overridden when the user actually types the command that resumes execution.

Reply:
> Okay so there might be a misunderstanding here. The sentinel command (step) will not "resume execution" of the program. 


I get that. Where is the misunderstanding?

Reply:
> The `step` command we append will only run `set_step()` and break out of the cmdloop - that's all we want. We need to get user input now, like the `step` command never executed. However, `set_step()` has side effect - it will write to some variables (to control which line the debugger needs to stop at when it yields control). But those side effects will be overridden when the user actually types the command that resumes execution.

Why not create a real sentinel then, instead of using step?

Reply:
In your modification you changed the comment to "the last command in the queue resumed execution" so I thought there might be some misunderstanding. Maybe it's just the wording.

How to create a real sentinel? I thought of that, and the direct way is to add a new debugger command that sets nothing and return `1` to exit the cmdloop. That seems even more messy because it adds a command which could appear in the auto-completion.

Reply:
> In your modification you changed the comment to "the last command in the queue resumed execution" so I thought there might be some misunderstanding. Maybe it's just the wording.
> 

Should have said last before the sentinel.

> 
> 
> How to create a real sentinel? I thought of that, and the direct way is to add a new debugger command that sets nothing and return `1` to exit the cmdloop. That seems even more messy because it adds a command which could appear in the auto-completion.

You can call it nop.

Reply:
Or 'pass' rather than 'nop'.

Reply:
> Should have said last before the sentinel.

If the sentinel is consumed, it means the existing command does *not* resume the execution. And neither should the sentinel - the sentinel is there to break out of the cmdloop.

`pass` is a bad command name because it's a reserved word. We had enonugh issue for `list` and `break`. However, `nop`, or `pass`, they don't describe the command. The sentinel command is not a no-op, it will repeat the last resume command (`continue`, `step`, `next` etc). It's a dangerous command because if the user uses it, they will resume the program being debugged. If we create a sentinel command, it should be internal only because it will have a very indescribable effect. That's why I simply chose `step` - it will have the exact same effect as a new command because under the current structure of `pdb` there's no way the side effect matters.

Reply:
I don't think 'sentinel' is the right name for that either.

A bugfix should not be so hacky and hard to explain, in my opinion.  Maybe it's me - if another core dev understands this and is happy to approve this PR as is I won't object to that.